src/Core/Framework/Adapter/Cache/InvalidatorStorage/RedisInvalidatorStorage.php

```diff
@@ -27,14 +27,27 @@ public function store(array $tags): void
 
     public function loadAndDelete(): array
     {
-        /** @var array{0: list<string>, 1: mixed} $values */
-        $values = $this
-            ->redis
-            ->multi()
-            ->sMembers(self::KEY)
-            ->del(self::KEY)
-            ->exec();
-
-        return $values[0];
+        try {
+            /** @var array{0: list<string>, 1: mixed}|false $values */
+            $values = $this
+                ->redis
+                ->multi()
+                ->sMembers(self::KEY)
+                ->del(self::KEY)
+                ->exec();
+
+            if ($values !== false) {
+                return $values[0];
+            }
+        } catch (\Throwable) {
+            // If the transaction fails (e.g. OOM), we fall back to sequential execution
+        }
+
+        // This breaks atomicity but ensures the queue is drained
+        /** @var list<string> $tags */
+        $tags = $this->redis->sMembers(self::KEY);
```

Comment:
should we batch the keys here?

Reply:
That would increase complexity, and basically, we are in a fallback scenario. 
So, I would say no, but yes, we should add a logger.

Reply:
I just wondered, if that could also cause OOM (again), if we send many many keys :man_shrugging: